refactor: Modify audio codec API to use AudioBuffer

yezhan10 opened the pull request:
What's included in this PR:
1. All audio encoders consume AudioBuffer, return AudioFrame pointer. 
2. All audio decoders consume AudioFrame, return AudioBuffer shared pointer obj. Since it's possible that a audio frame has multiple decoded PCM data. So split the original Decode() function to Decode() and GetDecodedAudioFrame(). 
3. Added unit tests for AudioPipe, OpusEncoder. 

Timestamp improvement will be included in next PR.

murillo128 commented on CMakeLists.txt:
```diff
@@ -56,14 +56,14 @@ add_library(MediaServerLib
-    ${SOURCES_G711}
-    ${SOURCES_G722}
-    ${SOURCES_GSM}
+    # ${SOURCES_MP3}
```
please revert

murillo128 commented on CMakeLists.txt:
```diff
@@ -126,8 +126,9 @@ add_library(MediaServerLib
     ${CMAKE_CURRENT_LIST_DIR}/src/VideoDecoderWorker.cpp
-    ${CMAKE_CURRENT_LIST_DIR}/src/audiodecoder.cpp
-    ${CMAKE_CURRENT_LIST_DIR}/src/audiotransrater.cpp
+    ${CMAKE_CURRENT_LIST_DIR}/src/AudioDecoderWorker.cpp
```
🥳

murillo128 commented on include/AudioPipe.h:
```diff
@@ -15,14 +17,14 @@ class AudioPipe :
 	
 	//Audio input
-	virtual int RecBuffer(SWORD *buffer,DWORD size);
+	virtual DWORD RecBuffer(DWORD frameSize, const AudioBuffer::shared& audioBuffer);
```
Let's return it instead
```suggestion
	virtual  AudioBuffer::shared RecBuffer(DWORD frameSize);
```

yezhan10 commented on include/AudioPipe.h:
```diff
@@ -15,14 +17,14 @@ class AudioPipe :
 
 	//Audio output
-	virtual int PlayBuffer(SWORD *buffer,DWORD size,DWORD frameTime, BYTE vadLevel = -1);
+	virtual int PlayBuffer(SWORD *buffer,DWORD size, QWORD pts, BYTE vadLevel = -1);
```
My initial thought is to push audio buffer to a queue. I discarded this since decoded audio frame size and required audio frame size may differ, for aac, each playbuffer we write 1024*numChannels worth of samples, but opus requires totally different frame size (960, 1920 etc) to encode. so the logic of consuming the pcm data might be complicated, and i still need a small fifo buffer serves as a cache in case requested number of samples from encoder is less than number of sampling in audio buffer.

yezhan10 commented on include/AudioPipe.h:
```diff
@@ -38,6 +40,14 @@ class AudioPipe :
 
 	//Members
 	fifo<SWORD,48000*4>	fifoBuffer;
+	std::queue<std::pair<QWORD, QWORD>> audioPts;
```
i think it might be more complicated if we have a queue of audio buffers? I will give more thoughts on this..

bcostdolby commented on src/AudioPipe.cpp:
```diff
@@ -262,33 +212,156 @@ int AudioPipe::RecBuffer(SWORD* buffer, DWORD size)
+	// if timestamp goes backwards
+	if(reset) 
+	{
+		if(encoderPTS + frameSize <= ref + decoderFrameSize)
```
Please rename everything to be consistent. Its difficult to read as every time i find a new variable I need to trace back how it is used to know what it is.

I think (are these meanings correct?):
* encoderNextPTS = encoderPTS
* encoderCurrentPTS = ts
* encodeFrameSize = frameSize 
* decoderNextPTS = ref
* decoderFrameSize = decoderFrameSize
* decoderFirst = first

bcostdolby replied:
I may have been very wrong in the above...

It looks like encoderPTS at the start of this function call holds  the PTS expected to be given to the first sample in the next block we pull from the fifo?

bcostdolby commented on src/AudioPipe.cpp:
```diff
@@ -262,33 +212,156 @@ int AudioPipe::RecBuffer(SWORD* buffer, DWORD size)
+		else
+		{
+			// no need to remove any data from fifoBuffer in case of time wraparound
+			// since timestamps are considered as continuous
```
But didnt we just maybe drop multiple items from the audioPts queue above?

bcostdolby commented on src/AudioPipe.cpp:
```diff
@@ -262,33 +212,156 @@ int AudioPipe::RecBuffer(SWORD* buffer, DWORD size)
+			// if decoder timestamp jump more than 2 frames
+			if(audioPts.front().first - ref > 2*decoderFrameSize)
+			{
+				if(encoderPTS + frameSize <= ref + decoderFrameSize)
```
Think I need to ask about this as well

murillo128 commented:
In order to be able to progress this PR, can we split it in two PRs? 

The first one about method changes to use the AudioBuffer and the second one to fix the timestamp calculation inside the AudioPipe?

yezhan10 commented:
> In order to be able to progress this PR, can we split it in two PRs?
> 
> The first one about method changes to use the AudioBuffer and the second one to fix the timestamp calculation inside the AudioPipe?

@murillo128 @bcostdolby I have updated the implementation to use a queue of audio buffers in audio pipe, and also added timestamp calculation for audio buffer based on decoder pts.  if we all like this approach better, i will clean up the code a bit in my next commit. This version still has `Encode(const AudioBuffer::const_shared& audioBuffer, BYTE* out, int outLen)`, I will update this to return an encoded frame in next PR most likely. Also, maybe we can use audiobuffer pool as an improvement in the next PR as well to avoid frequent audio buffer creation.

bcostdolby commented on include/AudioBuffer.h:
```diff
@@ -0,0 +1,109 @@
+		if(!pcmData || !*pcmData)
+			return Error("-AudioBuffer::CopyDecodedData() invalid frame data pointer\n");
+	
+		if(numSamples*numChannels > pcmBuffer.capacity()) 
```
Probably should be looking at size not capacity unless we are resizing the buffer as we go.

Also the each channel is doing a std::min with 2 and this does not so will miss the check

bcostdolby commented on include/AudioBuffer.h:
```diff
@@ -0,0 +1,109 @@
+
+	int CopyDecodedData(uint8_t** pcmData, uint32_t numSamples)
+	{
+		if(!pcmData || !*pcmData)
```
Possibly need to check more than just pcmData[0] != nullptr but also for all other possible channels. 

Overall this API isnt very safe might be worth rethinking. We have num samples passed in but not num channels and it is implicitly expected to be equal to min(numChannels, 2) which is non obvious

yezhan10 replied:
the min(numChannels,2) is from the old code, i guess this indicates that our system does not support multichannel audio.  i have removed the min(numChannels,2) and added check for each channel's data pointer

bcostdolby commented on src/AudioPipe.cpp:
```diff
@@ -156,78 +154,65 @@ int AudioPipe::StopPlaying()
-		fifoBuffer.remove(totalSize - static_cast<DWORD>(left / numChannels)* numChannels);
+		QWORD decoderPTS = audioBuffer->GetTimestamp();
+		// scale the decoder pts based on record rate
+		QWORD scaledDecoderPTS = std::round<QWORD>((decoderPTS-decoderPTSOffset) * (recordRate / (double)playRate)) + decoderPTSOffset;
```
Tomorrow when we chat again, might need to understand why we need to scale the PTS based on the resampling. I.e. For example if the ingest clock rate as from SRT is say 1000Hz and the PTS is 5000 (so 5 sec in roughly) and the incoming audio sample rate is say 16KHz (which is different from the PTS clock).

Lets say we need to resample to 32KHz we still would push PTS meta-data for the first sample in the block with a PTS = 5000 and a PTS Clock = 1000Hz as although there are now twice as many samples, they are still for the same point in time according to the PTS clock.

I dont see us changing the PTS clock in the code here (dont think we need to, but to work with this rescaling of PTS I assume we would also need to rescale the PTS clock from 1000Hz to 2000Hz).

bcostdolby commented on src/AudioPipe.cpp:
```diff
@@ -239,21 +224,21 @@ int AudioPipe::RecBuffer(SWORD* buffer, DWORD size)
-	while (!canceled && recording && (fifoBuffer.length() < totalSize + cache))
+	// TODO: use object pool to reuse audiobuffer
+	AudioBuffer::shared audioBuffer = std::make_shared<AudioBuffer>(frameSize, numChannels);
+	audioBuffer->SetClockRate(recordRate);
```
Ahh related to comment above. I see we are changing the PTS clock here. I dont think we need to do this right?

I.e. The sample rate clock and the PTS clock are two different things arent they?

bcostdolby commented on src/AudioPipe.cpp:
```diff
@@ -262,33 +247,117 @@ int AudioPipe::RecBuffer(SWORD* buffer, DWORD size)
+	// if rateBlockBuffer has enough data, no need to pop the audio buffer queue
+	if(cachedSamples >= frameSize)
+	{
+		rateBlockBuffer.pop(audioBuffer->GetAudioBufferData(requiredNumSamples), requiredNumSamples);
```
Dont need separate if statement block for this. Just take min(requiredNumSamples,cachedSamples) from cache (might be 0 or more up to needed size)

I.e. I think most of these if/else blocks can be flattened into a single loop.

yezhan10 commented on include/AudioBuffer.h:
```diff
@@ -0,0 +1,109 @@
+
+	void Resize(int numSamples) 
+	{
+		pcmBuffer.resize(numSamples*numChannels);
```
i removed this function since it's unused

murillo128 commented on include/AudioBuffer.h:
```diff
@@ -0,0 +1,114 @@
+		}
+		return totalWrittenSamples / numChannels;		
+	}
+	bool AddPCM(SWORD val, int pos) 
```
I am not sure if I prefer having a getter for the `std::vector` or overloading the  `[]` instead. 

At most
```suggestion
	bool SetSample(size_t pos, SWORD sample) 
```

yezhan10 replied:
I chose to use getter function and i would rename it to SetSampleAt() to have more obvious distinction between SetSamples()

murillo128 commented on include/AudioBuffer.h:
```diff
@@ -0,0 +1,114 @@
+	QWORD ts = 0;
+	DWORD clockRate = 0;
+	std::vector<SWORD> pcmBuffer;
+	SWORD* data;
```
I don't think we should keep reader state inside this class

yezhan10 replied:
the reader position is needed in ConsumeSamples() function which is used in gsmcodec.cpp. I will leave it for now.

murillo128 commented on include/AudioPipe.h:
```diff
@@ -34,10 +37,11 @@ class AudioPipe :
 	//Members
-	fifo<SWORD,48000*4>	fifoBuffer;
+	fifo<SWORD,48000*4>	rateBlockBuffer;
+	QWORD encoderPTS = 0;
```
let's remove for now

murillo128 commented on src/opus/OpusEncoder.h:
```diff
@@ -16,13 +17,16 @@ class OpusEncoder : public AudioEncoder
 	OpusEncoder(const Properties &properties);
 	virtual ~OpusEncoder();
-	virtual int Encode(SWORD *in,int inLen,BYTE* out,int outLen);
+	virtual AudioFrame* Encode(const AudioBuffer::const_shared& audioBuffer);
```
can we return a AudioFrame::shared better? We could add an <ObjectPool> to the  encoders for increasing performance if needed.

murillo128 commented on include/audio.h:
```diff
@@ -83,7 +84,7 @@ class AudioEncoder
 	// Must have virtual destructor to ensure child class's destructor is called
 	virtual ~AudioEncoder(){};
-	virtual int   Encode(SWORD *in,int inLen,BYTE* out,int outLen)=0;
+	virtual AudioFrame* Encode(const AudioBuffer::const_shared& audioBuffer) = 0;
```
can we use a shared pointer better?### 
```suggestion
	virtual AudioFrame::shared Encode(const AudioBuffer::const_shared& audioBuffer) = 0;

```

murillo128 reviewed (changes requested):
looking better, still some changes I would like to make

yezhan10 commented on include/AudioBuffer.h:
```diff
@@ -0,0 +1,114 @@
+	void	SetClockRate(DWORD clockRate) { this->clockRate = clockRate; }
+
+	const SWORD* GetData() const { return pcmBuffer.data(); }
+	SWORD* ConsumeData(int numSamples) 
```
this function is used in gsmcodec.cpp gsm_decode() line 92/96.

murillo128 replied:
I don't like keeping a read state inside the `AudioBuffer`. The same `AudioBuffer` could be passed to multiple `AudioEncoders` so it feels weird do modify the state just when reading samples.

murillo128 commented on src/gsm/gsmcodec.cpp:
```diff
@@ -51,42 +55,49 @@ GSMDecoder::~GSMDecoder()
 		//Decodificamos el primero
-		if (gsm_decode(g,(gsm_byte *)in,(gsm_signal *)out)<0)
-			return 0;
+		if (gsm_decode(g,(gsm_byte *)in,(gsm_signal *)audioBuffer->ConsumeSamples(160))<0)
```
```suggestion
		if (gsm_decode(g,(gsm_byte *)in,(gsm_signal *)audioBuffer->GetData(160))<0)
```

murillo128 commented on src/gsm/gsmcodec.cpp:
```diff
@@ -51,42 +55,49 @@ GSMDecoder::~GSMDecoder()
 		//Y el segundo
-		if (gsm_decode(g,(gsm_byte *)&in[33],(gsm_signal *)&out[160])<0)
-			return 0;
+		if (gsm_decode(g,(gsm_byte *)&in[33],(gsm_signal *)audioBuffer->ConsumeSamples(160))<0)
```
```suggestion
		if (gsm_decode(g,(gsm_byte *)&in[33],(gsm_signal *)(audioBuffer->GetData()+160))<0)
```

murillo128 commented on src/gsm/gsmcodec.cpp:
```diff
@@ -51,42 +55,49 @@ GSMDecoder::~GSMDecoder()
+	uint8_t* in =  const_cast<uint8_t*>(audioFrame->GetData());
+	if(!in || inLen==0) 
+		return 0;
+	audioFrameInfo.first = in;
```
why don't acually decode here and have a FIFO of the produced `AudioBuffers` that are consumed in the `GetDecodedAudioFrame()`?

As you are not keeping the `audioFrame` shared pointer, `in` pointer can be invalid when `GetDecodedAudioFrame()` is called

yezhan10 replied:
i agree that decode here instead of in GetDecodeAudioFrame() is more safe. Do we need a queue of audiobuffers stored for these codecs(g711 etc)? Looks like audioFrame only contains one decoded frame for these codecs.

murillo128 commented on src/g711/pcmacodec.cpp:
```diff
@@ -25,31 +26,49 @@ PCMADecoder::~PCMADecoder()
+
+AudioBuffer::shared PCMADecoder::GetDecodedAudioFrame()
+{
+	if(!audioBufferQueue.empty())
```
not big deal, but I preffer to fail early

```
if(audioBufferQueue.empty())
  return {};
auto audioBuffer = audioBufferQueue.front();
audioBufferQueue.pop();
return audioBuffer;
}

```

murillo128 reviewed (approved):
lgtm
